src/lm/sentence.cpp

```diff
@@ -0,0 +1,177 @@
+/**
+ * @file sentence.cpp
+ * @author Sean Massung
+ */
+
+#include <iostream>
+#include <sstream>
+#include <numeric>
+#include <iterator>
+#include "lm/sentence.h"
+#include "analyzers/analyzer.h"
+#include "analyzers/tokenizers/icu_tokenizer.h"
+#include "analyzers/tokenizers/whitespace_tokenizer.h"
+#include "analyzers/filters/all.h"
+
+namespace meta
+{
+namespace lm
+{
+sentence::sentence(const std::string& text, bool tokenize /* = true */)
+{
+    if (tokenize)
+    {
+        using namespace analyzers;
+        std::unique_ptr<token_stream> stream;
+        stream = make_unique<tokenizers::icu_tokenizer>();
+        stream = make_unique<filters::empty_sentence_filter>(std::move(stream));
+        std::string text_copy{text}; // consider changing parameter to non-const
+        stream->set_content(std::move(text_copy));
+        while (*stream)
+            tokens_.push_back(stream->next());
+
+        if (tokens_.empty())
+            throw sentence_exception{"empty token stream"};
+
+        // remove sentence markers
+        tokens_.pop_front();
+        tokens_.pop_back();
+
+        if (tokens_.empty())
+            throw sentence_exception{"empty token stream"};
+    }
+    else
+    {
+        std::istringstream iss{text};
+        std::copy(std::istream_iterator<std::string>(iss),
+                  std::istream_iterator<std::string>(),
+                  std::back_inserter(tokens_));
+    }
+}
+
+std::string sentence::to_string() const
+{
+    std::string result{""};
+    if (tokens_.empty())
+        return result;
+
+    for (auto& token : tokens_)
+        result += token + " ";
+
+    return result.substr(0, result.size() - 1); // remove trailing space
```

Comment:
All this string copying is probably really bad for performance. I think this class is a perfect optimization point for your `lm::diff` stuff. Here's what I'm thinking:

Instead of having your hash tables/sets having `std::string` as their keys (which you're comparing against some `temp_sentence.to_string()`), just use `sentence` objects directly. Compute the hash code of a `sentence` via a hash combine of all of its individual tokens' hash value (easy with `util::murmur_hash<>`), and equality comparison is just "do these have the same tokens in the same places". Then, when you've got a candidate `sentence` you want to check that you haven't seen yet, you can just look it up in your table directly and forgo the conversion to `std::string`.